### ISSUE TITLE
When chapters have subchapters

### ISSUE THREAD

**Issue by gerhaher** [koreader/koreader#5500]
#### Issue
When a chapter has subchapters the Mini progress bar has chapter markers only for the subchapters, but not for the chapter itself. 
For example in the book below there is a marker for 'Making everything gay' but not for '1 Gay'. 
And when using 'Next chapter'-gesture you jump directly from 'Introduction' to 'Making everything gay'.

The Full progress bar works the other way around: it leaves out the subchapters and instead has chapter markers for "1 Gay", "Interlude", "2 Women" etc.

Is this behaviour intentional? Would it be possible to change so that it takes account of both (chapter & subchapters).
Usually this is a nonissue because the chapter and its first subchapter are on the same page. But not so with this book as there is a lot of text before the first subchapters appear. 

![Reader_2019-Oct-17_140041](https://user-images.githubusercontent.com/32645430/67008442-c1bdd300-f0e9-11e9-84fa-bcc3bc454b44.png)
![Reader_2019-Oct-17_140111](https://user-images.githubusercontent.com/32645430/67008445-c5515a00-f0e9-11e9-94d2-71ecc5a5e828.png)

**Comment by NiLuJe** [comment]
How does the Skim widget deal with this?

(I *think* it does both, but, as you said, it's usually a non-issue because of the same-page-iness ;p).

**Comment by gerhaher** [comment]
> How does the Skim widget deal with this?
> 
> (I think it does both, but, as you said, it's usually a non-issue because of the same-page-iness ;p).

No, just the subchapter. For example, it jumps from 'Introduction' to 'Making Everything Gay'.

**Comment by poire-z** [comment]
> Is this behaviour intentional?

I have no idea. It's like this for years, and it's kinda explicite (even if no comment...) to behave in this strange way: walk the TOC from the end, and ignore TOC entries with a lower level than one we have met:
https://github.com/koreader/koreader/blob/7de022a61742f42224a046b604202e3a8e150ed7/frontend/apps/reader/modules/readertoc.lua#L142-L158

Looking around, it's like this since https://github.com/koreader/koreader/pull/984. Explanation at https://github.com/koreader/koreader/issues/976#issuecomment-58483192.

Personnally, since a few years, I fix that in my own patch by just commenting line 153:
```diff
@@ -153,7 +153,7 @@
                 if v.depth >= depth then
                     table.insert(ticks, v.page)
                 end
-                depth = v.depth
+                -- depth = v.depth -- so we get them all and don't ignore toc entries that have children
             end
         else
             local depth
```
No issue for me, may be a sometimes crowded bottom progress bar, but may be it'd be the same without that patch.

**Comment by gerhaher** [comment]
> Personnally, since a few years, I fix that in my own patch by just commenting line 156:

Thanks, I prefer this too.

**Comment by poire-z** [comment]
Another complain (and good user investigation that came up to the right conclusion :):
https://www.mobileread.com/forums/showthread.php?t=333417

Original reasoning for that at https://github.com/koreader/koreader/issues/976#issuecomment-58483192 (I don't want to re-read it, so please look at it to see if it really makes sense).

Now that chapters can cause a full flashing refresh, may be we should go the way of adding a tick for all chapters, and not only leaf ones?
(One line to comment out, see https://github.com/koreader/koreader/issues/5500#issuecomment-543201699 - and see at the other stuff like SkimTo so everything show the same ticks.)

(It was one of the earliest fix I did for myself, it's strange users don't complain much more than that, I've seen no more than 4 or 5 - so, dunno :)

**Comment by NiLuJe** [comment]
I definitely agree with you, I'll look at the issue you linked, thanks ;).

**Comment by NiLuJe** [comment]
Okay, the idea was to somehow keep track of the nesting level in the tick table.

Thing is, I don't think anything actually cares about this: everything seems to pass level 0 to the tick stuff, so, while keeping the hierarchy intact would be NTH, it's not trivial to get right, and we don't make proper use of it anyway.

So, perhaps it could be massively simplified and just tick All The Things and ditch the level tracking there?

(I mean, it makes sense to honor level when displaying the actual ToC, but for the tick themselves, eh).

**Comment by NiLuJe** [comment]
Hmm, SkimToWidget does something slightly neat with the level tracking, though (I think? :D): it *only* ticks the deepest level...

Which, okay, avoids putting three ticks on one page when you have I.i.1 headers in basically the same place...

Hmm.

**Comment by NiLuJe** [comment]
Then again, that could be handled another way: simply drop duplicate ticks.

**Comment by NiLuJe** [comment]
Okay, re-re-reading https://github.com/koreader/koreader/issues/976#issuecomment-58483192, SkimToWidget actually implements the `{i, ii, iii, iv, vi, v}` variant.

I was mistakenly under the impression that the issue was level tracking at the earliest level (i.e., finding out in which level to sort a tick in), but, no, it's finding out *which* bits to keep based on that hierarchical table ^^.

TL;DR: In which case, yeah, I think keeping all of 'em is probably the best way forward, possibly dropping duplicates, at the cost of sometimes having crowded nests of ticks when subsections are started close to each other?

**Comment by NiLuJe** [comment]
... and https://github.com/koreader/koreader/pull/984 made level 0 the "find all leaf nodes, even across levels" approach, which explains the confusion, since that's what we mostly use ^^.

**Comment by NiLuJe** [comment]
Does anyone has a good ready-to-use testcase with wonky and heavy nesting?

**Comment by poire-z** [comment]
> Does anyone have a good ready-to-use testcase with wonky and heavy nesting?

Wikipedia EPUBs have some nesting on a same page (not heavy heavy, but easier to grasp). It's what you tested chapter flashes issues with.

(I still don't want to re-read and think hierarchy and leaf etc... so not following all your thoughts :) 

> In which case, yeah, I think keeping all of 'em is probably the best way forward, possibly dropping duplicates, at the cost of sometimes having crowded nests of ticks when subsections are started close to each other?

Duplicates are not bothering, as you may just be drawing the same line at the same place (but ok, may be a performance improvement by not doing what's unneeded) - and the "get next tick/chapter" stuff must already skip ticks that are on the same page (as we always get to another page).
As for the crowded nest of ticks, that's what I'm leaving with :) and I'm fine with it. Sometimes it's too crowded to see anything, but I can guess the progression from the density / grey level :) - but this could happen even with leafs. But I least, I don't miss any chapter when it's not crowded (which was why I made that fix on my side: I would see a bit part of blank, and suddently meet a chapter where there was no tick indicating it!).
